# Seed gene discovery

This pipeline discovers *seed genes* for DeepRVAT training. The pipeline runs SKAT and burden tests for missense and pLOF variants, weighting variants with Beta(MAF,1,25). To run the tests, we use the `Scoretest` from the [SEAK](https://github.com/HealthML/seak) package (has to be installed from github).

To run the pipeline, an experiment directory with the `config.yaml` has to be created. An `lsf.yaml` file specifiying the compute resources for each rule in `seed_gene_discovery.snakefile` might also be needed depending on your system (see as an example the `lsf.yaml` file in this directory). 

## Input data

The experiment directory in addition requires to have the same input data as specified for [DeepRVAT](deeprvat.md), including
- `annotations.parquet`
- `protein_coding_genes.parquet`
- `genotypes.h5`
- `variants.parquet`
- `phenotypes.parquet`

The `annotations.parquet` data frame output by the annotation pipeline can be used (although only the `Consequence_missense_variant` and `MAF` column will be used). One has to add a column `is_plof` to indicate whether a variant should is a loss of function variant (1) or not (0). We recommend to set this to `1` if any of the VEP consequences  `["splice_acceptor_variant", "splice_donor_variant",  "frameshift_variant", "stop_gained", "stop_lost", "start_lost"]`


The `annotations.parquet` dataframe, generated by the annotation pipeline, can be utilized. To indicate if a variant is a loss of function (pLoF) variant, a column `is_plof` has to be added with values 0 or 1. We recommend to set this to `1` if the variant has been classified as any of these VEP consequences `["splice_acceptor_variant", "splice_donor_variant", "frameshift_variant", "stop_gained", "stop_lost", "start_lost"]`.

## Configuration file

*TODO:* Describe `config.yaml`, give example


## Running the seed gene discovery pipeline

In a directory with all of the [input data](#input-data) required and your [configuration file](#configuration-file) set up, run: 

```
[path_to_deeprvat]/pipelines/seed_gene_discovery.snakefile
```

Replace `[path_to_deeprvat]` with the path to your clone of the repository.

